pkg/trace/sampler/coresampler.go

```diff
@@ -41,14 +41,20 @@ const (
 	PriorityEngineType
 )
 
+type SampleDecision struct {
```

Comment:
I don't think we need to make any changes in this package nor in `pkg/agent/sampler.go`. IMHO we should leave this package as it is, in a simpler way and make all the changes in `pkg/trace/agent/agent.go`, having `runSamplers` return more details. Inside `runSamplers` you know which sampler decided what, no need to change the interface, the samplers themselves, and add new exported identifiers.

All we need is to know which sampler took the decision, I'm thinking something like:

```diff
diff --git a/pkg/trace/agent/agent.go b/pkg/trace/agent/agent.go
index bff19caaa..d303d97bf 100644
--- a/pkg/trace/agent/agent.go
+++ b/pkg/trace/agent/agent.go
@@ -7,6 +7,7 @@ package agent
 
 import (
 	"context"
+	"fmt"
 	"runtime"
 	"sync/atomic"
 	"time"
@@ -16,6 +17,7 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/trace/event"
 	"github.com/DataDog/datadog-agent/pkg/trace/filters"
 	"github.com/DataDog/datadog-agent/pkg/trace/info"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
 	"github.com/DataDog/datadog-agent/pkg/trace/metrics/timing"
 	"github.com/DataDog/datadog-agent/pkg/trace/obfuscate"
 	"github.com/DataDog/datadog-agent/pkg/trace/pb"
@@ -246,10 +248,17 @@ func (a *Agent) Process(t *api.Trace) {
 func (a *Agent) sample(ts *info.TagStats, pt ProcessedTrace) {
 	var ss writer.SampledSpans
 
-	sampled, rate := a.runSamplers(pt)
+	sampled, rate, which := a.runSamplers(pt)
 	if sampled {
 		sampler.AddGlobalRate(pt.Root, rate)
 		ss.Trace = pt.Trace
+		metrics.Count("datadog.trace_agent.sampling", 1, []string{
+			"decision:keep",
+			fmt.Sprintf("service:%s", pt.Root.Service),
+			fmt.Sprintf("sampler:%s", which),
+		}, 1)
+	} else {
+		metrics.Count("datadog.trace_agent.sampling", 1, []string{"decision:drop", fmt.Sprintf("service:%s", pt.Root.Service)}, 1)
 	}
 
 	events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
@@ -265,7 +274,7 @@ func (a *Agent) sample(ts *info.TagStats, pt ProcessedTrace) {
 
 // runSamplers runs all the agent's samplers on pt and returns the sampling decision
 // along with the sampling rate.
-func (a *Agent) runSamplers(pt ProcessedTrace) (bool, float64) {
+func (a *Agent) runSamplers(pt ProcessedTrace) (sampled bool, rate float64, which string) {
 	if _, ok := pt.GetSamplingPriority(); ok {
 		return a.samplePriorityTrace(pt)
 	}
@@ -275,25 +284,33 @@ func (a *Agent) runSamplers(pt ProcessedTrace) (bool, float64) {
 // samplePriorityTrace samples traces with priority set on them. PrioritySampler and
 // ErrorSampler are run in parallel. The ExceptionSampler catches traces with rare top-level
 // or measured spans that are not caught by PrioritySampler and ErrorSampler.
-func (a *Agent) samplePriorityTrace(pt ProcessedTrace) (sampled bool, rate float64) {
+func (a *Agent) samplePriorityTrace(pt ProcessedTrace) (sampled bool, rate float64, which string) {
 	sampledPriority, ratePriority := a.PrioritySampler.Add(pt)
+	if sampledPriority {
+		which = "priority"
+	}
 	if traceContainsError(pt.Trace) {
 		sampledError, rateError := a.ErrorsScoreSampler.Add(pt)
-		return sampledError || sampledPriority, sampler.CombineRates(ratePriority, rateError)
+		if !sampledPriority && sampledError {
+			which = "error"
+		}
+		return sampledError || sampledPriority, sampler.CombineRates(ratePriority, rateError), which
 	}
 	if sampled := a.ExceptionSampler.Add(pt.Env, pt.Root, pt.Trace); sampled {
-		return sampled, 1
+		return sampled, 1, "exception"
 	}
-	return sampledPriority, ratePriority
+	return sampledPriority, ratePriority, which
 }
 
 // sampleNoPriorityTrace samples traces with no priority set on them. The traces
 // get sampled by either the score sampler or the error sampler if they have an error.
-func (a *Agent) sampleNoPriorityTrace(pt ProcessedTrace) (sampled bool, rate float64) {
+func (a *Agent) sampleNoPriorityTrace(pt ProcessedTrace) (sampled bool, rate float64, which string) {
 	if traceContainsError(pt.Trace) {
-		return a.ErrorsScoreSampler.Add(pt)
+		sampled, rate = a.ErrorsScoreSampler.Add(pt)
+		return sampled, rate, "error"
 	}
-	return a.ScoreSampler.Add(pt)
+	sampled, rate = a.ScoreSampler.Add(pt)
+	return sampled, rate, "score"
 }
 
 func traceContainsError(trace pb.Trace) bool {
```

It would be nice if `which` would not be a `string`, but maybe its own type. Maybe `sampler.EngineType` is a good candidate (we'd have to add `Exception` to the list), although that type itself has the wrong reasoning behind it. Not sure.

What do you think?